--- crash-7.1.0/sadump.c.orig
+++ crash-7.1.0/sadump.c
@@ -394,7 +394,13 @@ restart:
 	}
 
 	sd->filename = file;
-	sd->flags = flags;
+
+	/*
+	 * Switch to zero excluded mode by default on sadump-related
+	 * formats because some Fujitsu troubleshooting software
+	 * assumes the behavior.
+	 */
+	sd->flags = flags | SADUMP_ZERO_EXCLUDED;
 
 	if (machine_type("X86"))
 		sd->machine_type = EM_386;
--- crash-7.1.0/tools.c.orig
+++ crash-7.1.0/tools.c
@@ -2479,7 +2479,9 @@ show_options(void)
 	fprintf(fp, "      namelist: %s\n", pc->namelist);
 	fprintf(fp, "      dumpfile: %s\n", pc->dumpfile);
 	fprintf(fp, "        unwind: %s\n", kt->flags & DWARF_UNWIND ? "on" : "off");
-	fprintf(fp, " zero_excluded: %s\n", *diskdump_flags & ZERO_EXCLUDED ? "on" : "off");
+	fprintf(fp, " zero_excluded: %s\n",
+		(*diskdump_flags & ZERO_EXCLUDED) || sadump_is_zero_excluded() ?
+		"on" : "off");
 	fprintf(fp, "     null-stop: %s\n", *gdb_stop_print_at_null ? "on" : "off");
 	fprintf(fp, "           gdb: %s\n", pc->flags2 & GDB_CMD_MODE ? "on" : "off");
 	fprintf(fp, "         scope: %lx ", pc->scope);
